apps/emqx/src/emqx_channel.erl

```diff
@@ -2854,6 +3091,108 @@ proto_ver(_Reason, #{proto_ver := ProtoVer}) ->
 proto_ver(_, _) ->
     ?MQTT_PROTO_V4.
 
+%%--------------------------------------------------------------------
+%% External Trace Helpers
```

Comment:
It's better to hide attrs building to emqx_external_trace.erl

Another advantage is that direct calls to the emqx_otel_tracer module can be avoided in the emqx_channel module.